packages/core/src/2d/text/TextRenderer.ts

```diff
@@ -0,0 +1,531 @@
+import { TextUtils } from "./TextUtils";
+
+export class TextRenderer extends Renderer {
+  private static _textureProperty: ShaderProperty = Shader.getPropertyByName("u_spriteTexture");
+  private static _tempVec2: Vector2 = new Vector2();
+  private static _tempVec3: Vector3 = new Vector3();
+
+  /** @internal temp solution. */
+  @ignoreClone
+  _customLocalBounds: BoundingBox = null;
+  /** @internal temp solution. */
+  @ignoreClone
+  _customRootEntity: Entity = null;
+
+  @assignmentClone
+  private _sprite: Sprite = null;
+  @deepClone
+  private _positions: Vector3[] = [new Vector3(), new Vector3(), new Vector3(), new Vector3()];
+  @deepClone
+  private _color: Color = new Color(1, 1, 1, 1);
+  @assignmentClone
+  private _text: string = "";
+  @assignmentClone
+  private _width: number = 0;
+  @assignmentClone
+  private _height: number = 0;
+  @assignmentClone
+  private _fontName: string = "Arial";
+  @assignmentClone
+  private _fontSize: number = 24;
+  @assignmentClone
+  private _lineSpace: number = 0;
+  @assignmentClone
+  private _isBold: boolean = false;
+  @assignmentClone
+  private _isItalic: boolean = false;
+  @assignmentClone
+  private _horizontalAlignment: TextHorizontalAlignment = TextHorizontalAlignment.Center;
+  @assignmentClone
+  private _verticalAlignment: TextVerticalAlignment = TextVerticalAlignment.Center;
+  @assignmentClone
+  private _horizontalOverflow: TextHorizontalOverflow = TextHorizontalOverflow.Overflow;
+  @assignmentClone
+  private _verticalOverflow: TextVerticalOverflow = TextVerticalOverflow.Overflow;
+  @ignoreClone
+  private _dirtyFlag: number = DirtyFlag.Style;
+  @ignoreClone
+  private _isWorldMatrixDirty: UpdateFlag;
+  @assignmentClone
+  private _maskInteraction: SpriteMaskInteraction = SpriteMaskInteraction.None;
+  @assignmentClone
+  private _maskLayer: number = SpriteMaskLayer.Layer0;
+
+  /**
+   * Rendering color for the TextRenderer.
+   */
+  get color(): Color {
+    return this._color;
+  }
+
+  set color(value: Color) {
+    if (this._color !== value) {
+      value.cloneTo(this._color);
+    }
+  }
+
+  /**
+   * Rendering string for the TextRenderer.
+   */
+  get text(): string {
+    return this._text;
+  }
+
+  set text(value: string) {
+    value = value || "";
+    if (this._text !== value) {
+      this._text = value;
+      this._setDirtyFlagTrue(DirtyFlag.Style);
+    }
+  }
+
+  /**
+   * The width of the TextRenderer (in 3D world coordinates).
+   */
+  get width(): number {
+    return this._width;
+  }
+
+  set width(value: number) {
+    if (this._width !== value) {
+      this._width = value;
+      this._setDirtyFlagTrue(DirtyFlag.Style);
+    }
+  }
+
+  /**
+   * The height of the TextRenderer (in 3D world coordinates).
+   */
+  get height(): number {
+    return this._height;
+  }
+
+  set height(value: number) {
+    if (this._height !== value) {
+      this._height = value;
+      this._setDirtyFlagTrue(DirtyFlag.Style);
+    }
+  }
```

Comment:
I understand why need width and height, later we will support `RectTransform`, we add width and height for temp solution now.